README.md
[![Gem version](http://img.shields.io/gem/v/rainpress.svg)](http://rubygems.org/gems/rainpress)
[![Build status](http://img.shields.io/travis/ddfreyne/rainpress.svg)](https://travis-ci.org/ddfreyne/rainpress)

# Rainpress

Rainpress is a compressor for CSS, written in Ruby. It removes unnecessary characters and replaces some attributes with a shorter equivalent name.

## Usage

```ruby
require 'rainpress'
compressed_style_text = Rainpress.compress(style_text)
```

Options:

* `:comments` — if set to false, comments will not be removed
* `:newlines` — if set to false, newlines will not be removed
* `:spaces` — if set to false, spaces will not be removed
* `:colors` — if set to false, colors will not be modified
* `:misc` — if set to false, miscellaneous compression parts will be skipped

## License

Copyright (c) 2007–2008 Uwe L. Korn

Copyright (c) 2009-2010 Jeff Smick

Copyright (c) 2017–… Denis Defreyne

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


NEWS.md
## 1.0.1

Enhancements:

* Improved compatibility with recent versions of Ruby, Rubygems and Bundler

## 1.0.0

Initial release.


spec/rainpress_spec.rb
require 'rainpress'
require 'minitest'
require 'minitest/spec'

Minitest.autorun

describe 'Rainpress module' do
  it 'removes comments' do
    options = {
      :comments => true,
      :newlines => false,
      :spaces   => false,
      :colors   => false,
      :misc     => false
    }

    # plain comment -> ''
    Rainpress.compress('/* sss */', options).must_equal ''

    # no comment -> no change
    Rainpress.compress('sss', options).must_equal 'sss'

    # comment floating in text
    Rainpress.compress('s/*ss*/ss', options).must_equal 'sss'

    # multiple comments floating in text
    Rainpress.compress('s/*ss*/ss/*ss*/s', options).must_equal 'ssss'

    # empty string
    Rainpress.compress('', options).must_equal ''
  end

  it 'removes newlines' do
    options = {
      :comments => false,
      :newlines => true,
      :spaces   => false,
      :colors   => false,
      :misc     => false
    }

    # plain unix-newline
    Rainpress.compress("\n", options).must_equal ''

    # plain mac newline
    Rainpress.compress("\r", options).must_equal ''

    # plain windows newline
    Rainpress.compress("\r\n", options).must_equal ''

    # no newline
    Rainpress.compress('rn', options).must_equal 'rn'

    # newlines floatin in text
    Rainpress.compress("sss\n||\r\nsss", options).must_equal 'sss||sss'

    # empty string
    Rainpress.compress('', options).must_equal ''
  end

  it 'removes spaces' do
    options = {
      :comments => false,
      :newlines => false,
      :spaces   => true,
      :colors   => false,
      :misc     => false
    }

    # (a) Turn mutiple Spaces into a single, but not less
    Rainpress.compress('  ', options).must_equal ' ' # 2 spaces
    Rainpress.compress('   ', options).must_equal ' ' # 3 spaces

    # (b) remove spaces around ;:{},
    Rainpress.compress(' ; ', options).must_equal ';'
    Rainpress.compress(' : ', options).must_equal ':'
    Rainpress.compress(' { ', options).must_equal '{'
    Rainpress.compress(' } ', options).must_equal '}'
    Rainpress.compress(',', options).must_equal ','

    # (c) remove tabs
    Rainpress.compress("\t", options).must_equal ''
  end

  it 'shortens colors' do
    options = {
      :comments => false,
      :newlines => false,
      :spaces   => false,
      :colors   => true,
      :misc     => false
    }
    # rgb(50,101,152) to #326598
    Rainpress.compress('color:rgb(12,101,152)', options).must_equal 'color:#0c6598'

    # #AABBCC to #ABC
    Rainpress.compress('color:#AAbBCC', options).must_equal 'color:#AbC'

    # Keep chroma(color="#FFFFFF"); ... due to IE
    Rainpress.compress('chroma(color="#FFFFFF");', options).must_equal 'chroma(color="#FFFFFF");'

    # shorten several names to numbers
    Rainpress.compress('color:white;', options).must_equal 'color:#fff;'
    Rainpress.compress('color: white}', options).must_equal 'color:#fff}'

    # shotern several numbers to names
    Rainpress.compress('color:#ff0000;', options).must_equal 'color:red;'
    Rainpress.compress('color:#F00;', options).must_equal 'color:red;'
  end

  it 'does misc' do
    options = {
      :comments => false,
      :newlines => false,
      :spaces   => false,
      :colors   => false,
      :misc     => true
    }
    # Replace 0(pt,px,em,%) with 0
    %w[px em pt % in cm mm pc ex].each { |x| Rainpress.compress(" 0#{x}", options).must_equal ' 0' }
    Rainpress.compress(':0mm', options).must_equal ':0'
    Rainpress.compress('  0ex', options).must_equal '  0'
    Rainpress.compress(' 10ex', options).must_equal ' 10ex'

    # Replace 0 0 0 0; with 0.
    Rainpress.compress(':0 0;', options).must_equal ':0;'
    Rainpress.compress(':0 0 0;', options).must_equal ':0;'
    Rainpress.compress(':0 0 0 0;', options).must_equal ':0;'
    Rainpress.compress(':0 0 0 0}', options).must_equal ':0}'

    # Keep 'background-position:0 0;' !!
    Rainpress.compress('background-position:0 0;', options).must_equal 'background-position:0 0;'

    # Replace 0.6 to .6, but only when preceded by : or a white-space
    Rainpress.compress(' 0.6', options).must_equal ' .6'
    Rainpress.compress(':0.06', options).must_equal ':.06'
    Rainpress.compress('10.6', options).must_equal '10.6'

    # Replace ;;;; with ;
    Rainpress.compress('ss;;;ss', options).must_equal 'ss;ss'

    # Replace ;} with }
    Rainpress.compress('ss;sss;}ss', options).must_equal 'ss;sss}ss'

    # Replace font-weight:normal; with 400, bold with 700
    Rainpress.compress('font-weight: normal;', options).must_equal 'font-weight:400;'
    Rainpress.compress('font: normal;', options).must_equal 'font: 400;'
    Rainpress.compress('font: bold 1px;', options).must_equal 'font: 700 1px;'
  end
end


lib/rainpress.rb
class Rainpress
  def self.compress(style, options = {})
    self.new(style, options).compress!
  end

  def initialize(style, opts = {})
    @style = style
    @opts = {
      :comments => true,
      :newlines => true,
      :spaces   => true,
      :colors   => true,
      :misc     => true
    }
    @opts.merge! opts
  end

  # Run the compressions and return the newly compressed text
  def compress!
    remove_comments!  if @opts[:comments]
    remove_newlines!  if @opts[:newlines]
    remove_spaces!    if @opts[:spaces]
    shorten_colors!   if @opts[:colors]
    do_misc!          if @opts[:misc]
    @style
  end

  # Remove all comments out of the CSS-Document
  #
  # Only /* text */ comments are supported.
  # Attention: If you are doing css hacks for IE using the comment tricks,
  # they will be removed using this function. Please consider for IE css style
  # corrections the usage of conditionals comments in your (X)HTML document.
  def remove_comments!
    input = @style
    @style = ''

    while input.length > 0 do
      pos = input.index("/*");

      # No more comments
      if pos == nil
        @style += input
        input = '';
      else # Comment beginning at pos
        @style += input[0..(pos-1)] if pos > 0 # only append text if there is some
        input = input[(pos+2)..-1]
        # Comment ending at pos
        pos = input.index("*/")
        input = input[(pos+2)..-1]
      end
    end
  end

  # Remove all newline characters
  #
  # We take care of Windows(\r\n), Unix(\n) and Mac(\r) newlines.
  def remove_newlines!
  @style.gsub!(/\n|\r/, '')
  end

  # Remove unneeded spaces
  #
  # 1. Turn mutiple spaces into a single
  # 2. Remove spaces around ;:{},
  # 3. Remove tabs
  def remove_spaces!
    @style.gsub!(/\s*(\s|;|:|\}|\{|,)\s*/, '\1')
    @style.gsub! "\t", ''
  end

  # Replace color values with their shorter equivalent
  #
  # 1. Turn rgb(,,)-colors into #-values
  # 2. Shorten #AABBCC down to #ABC
  # 3. Replace names with their shorter hex-equivalent
  #    * white -> #fff
 	#    * black -> #000
  # 4. Replace #-values with their shorter name
  #    * #f00 -> red
  def shorten_colors!
    # rgb(50,101,152) to #326598
    @style.gsub!(/rgb\s*\(\s*([0-9,\s]+)\s*\)/) do |match|
      out = '#'
      $1.split(',').each do |num|
        out += '0' if num.to_i < 16
        out += num.to_i.to_s(16) # convert to hex
      end
      out
    end
    # Convert #AABBCC to #ABC, keep if preceed by a '='
    @style.gsub!(/([^\"'=\s])(\s*)#([\da-f])\3([\da-f])\4([\da-f])\5/i, '\1#\3\4\5')

    # At the moment we assume that colours only appear before ';' or '}' and
    # after a ':', if there could be an occurence of a color before or after
    # an other character, submit either a bug report or, better, a patch that
    # enables Rainpress to take care of this.

    # shorten several names to numbers
    ## shorten white -> #fff
    @style.gsub!(/:\s*white\s*(;|\})/, ':#fff\1')

    ## shorten black -> #000
    @style.gsub!(/:\s*black\s*(;|\})/, ':#000\1')

    # shotern several numbers to names
    ## shorten #f00 or #ff0000 -> red
    @style.gsub!(/:\s*#f{1,2}0{2,4}(;|\})/i, ':red\1')
  end

  # Do miscellaneous compression methods on the style.
  def do_misc!
    # Replace 0(pt,px,em,%) with 0 but only when preceded by : or a white-space
    @style.gsub!(/([\s:]+)(0)(px|em|%|in|cm|mm|pc|pt|ex)/i, '\1\2')

    # Replace :0 0 0 0(;|}) with :0(;|})
    @style.gsub!(/:0 0 0 0(;|\})/, ':0\1')

    # Replace :0 0 0(;|}) with :0(;|})
    @style.gsub!(/:0 0 0(;|\})/, ':0\1')

    # Replace :0 0(;|}) with :0(;|})
    @style.gsub!(/:0 0(;|\})/, ':0\1')

    # Replace background-position:0; with background-position:0 0;
    @style.gsub! 'background-position:0;', 'background-position:0 0;'

    # Replace 0.6 to .6, but only when preceded by : or a white-space
    @style.gsub!(/[:\s]0+\.(\d+)/) do |match|
      match.sub '0', '' # only first '0' !!
    end

    # Replace multiple ';' with a single ';'
    @style.gsub!(/[;]+/, ';')

    # Replace ;} with }
    @style.gsub! ';}', '}'

    # Replace font-weight:normal; with 400
    @style.gsub!(/font-weight[\s]*:[\s]*normal[\s]*(;|\})/i,'font-weight:400\1')
    @style.gsub!(/font[\s]*:[\s]*normal[\s;\}]*/) do |match|
      match.sub 'normal', '400'
    end

    # Replace font-weight:bold; with 700
    @style.gsub!(/font-weight[\s]*:[\s]*bold[\s]*(;|\})/,'font-weight:700\1')
    @style.gsub!(/font[\s]*:[\s]*bold[\s;\}]*/) do |match|
      match.sub 'bold', '700'
    end
  end

end
